### ISSUE TITLE
Creating an Template Field item

### ISSUE THREAD

**Issue by mikeedwards83** [sergeyshushlyapin/Sitecore.FakeDb#80]
I am trying to create an item based on a template field:

	/sitecore/templates/System/Templates/Template field

My code looks like this:

    new DbItem("TestField", ID.NewID, new ID("{455A3E98-A627-4B40-8035-E683A0331AC7}"))

This causes the following exception:

    System.InvalidOperationException : Unable to create the item based on the template '{455A3E98-A627-4B40-8035-E683A0331AC7}'. An item with the same id has already been added ('/sitecore/templates/template field').
       at Sitecore.FakeDb.Data.Engines.DataStorage.AssertDoesNotExists(DbItem item)
       at Sitecore.FakeDb.Data.Engines.DataStorage.AddFakeItem(DbItem item)
       at Sitecore.FakeDb.Pipelines.AddDbItem.CreateTemplate.Process(AddDbItemArgs args)
       at (Object , Object[] )
       at Sitecore.Pipelines.PipelineMethod.Invoke(Object[] parameters)
       at Sitecore.Pipelines.CorePipeline.Run(PipelineArgs args)
       at Sitecore.Pipelines.CorePipeline.Run(String pipelineName, PipelineArgs args, String pipelineDomain, Boolean failIfNotExists)
       at Sitecore.Pipelines.CorePipeline.Run(String pipelineName, PipelineArgs args, String pipelineDomain)
       at Sitecore.Pipelines.CorePipeline.Run(String pipelineName, PipelineArgs args)
       at Sitecore.FakeDb.Data.Engines.DataStorage.AddFakeItem(DbItem item)
       at Sitecore.FakeDb.Data.Engines.DataStorage.AddFakeItem(DbItem item)
       at Sitecore.FakeDb.Data.Engines.DataStorage.AddFakeItem(DbItem item)
       at Sitecore.FakeDb.Data.Engines.DataStorage.AddFakeItem(DbItem item)
       at Sitecore.FakeDb.Db.Add(DbItem item)

Am I do something wrong?

**Comment by sergeyshushlyapin** [comment]
Hi Mike, 




That is the expected behavior. You're trying to create an item but the 'templateId' parameter contains not a **template** id. It contains the id of another **item** (which is already added to the database). 'Template field' is not a 'template'.



—

**Comment by sergeyshushlyapin** [comment]
Could you please add a bit more details to the scenario you want to achieve? Thanks.

**Comment by mikeedwards83** [comment]
I want to create an item based on the standard Template Field template in Sitecore because I have an On Save event that manipulates the Template Field to set a title and to make sure the field matches a naming convention.

**Comment by mikeedwards83** [comment]
I have been playing around a bit more and I notice that FakeDb doesn't create the template fields as items. Take for example this code:

 var tId1 = ID.NewID;
            var fId1 = ID.NewID;

            using (var db = new Db
            {                
                        new DbTemplate("TestTemplate1", tId1)
                        {
                            Fields = { new DbField("TestField",fId1)}
                        }
            })
            {


                Template temp = Sitecore.Data.Managers.TemplateManager.GetTemplate(tId1, db.Database);
                Assert.IsTrue(temp.ContainsField(fId1));  //this returns true

                var fieldItem = db.GetItem(fId1); 
                Assert.IsNotNull(fieldItem);  //this fails

            }


So the template gets the field but the actual field isn't create in the DB.

I need to be able to actually create the templates and fields in the DB and treat them as items.

**Comment by sergeyshushlyapin** [comment]
It is possible to *create* a template field item manually now, but the template fields are still not stored as items. I can rewrite your test in a slightly different way building the template using `DbItem`'s:

```csharp
using (var db = new Db
        {                
                    new DbTemplate("TestTemplate1", tId1)
                    {
                        new DbItem("TestField",fId1, TemplateIDs.TemplateField)
                    }
        })
      {
        var fieldItem = db.GetItem(fId1);
        Assert.NotNull(fieldItem);  // does not fail now

        Template temp = Sitecore.Data.Managers.TemplateManager.GetTemplate(tId1, db.Database);
        Assert.True(temp.ContainsField(fId1));  // NOW FAILS :)
      }
```

I think both scenarios should be supported. I'll keep working on this, thanks.

**Comment by sergeyshushlyapin** [comment]
Fixed and released as v0.29.1

**Comment by sergeyshushlyapin** [comment]
@mikeedwards83 please let me know if your issue has been solved. Thanks.
